temporal-sdk/src/main/java/io/temporal/internal/client/RootWorkflowClientInvoker.java

```diff
@@ -0,0 +1,191 @@
+  }
+
+  @Override
+  public <R> GetResultOutput<R> getResult(GetResultInput<R> input) throws TimeoutException {
+    Optional<Payloads> resultValue =
+        WorkflowExecutionUtils.getWorkflowExecutionResult(
+            genericClient.getService(),
+            genericClient.getNamespace(),
+            input.getWorkflowExecution(),
+            input.getWorkflowType(),
+            metricsScope,
+            clientOptions.getDataConverter(),
+            input.getTimeout(),
+            input.getTimeoutUnit());
+    return new GetResultOutput<>(
+        convertResultPayloads(resultValue, input.getResultClass(), input.getResultType()));
+  }
+
+  @Override
+  public <R> GetResultAsyncOutput<R> getResultAsync(GetResultInput<R> input) {
+    CompletableFuture<Optional<Payloads>> resultValue =
+        WorkflowExecutionUtils.getWorkflowExecutionResultAsync(
+            genericClient.getService(),
+            genericClient.getNamespace(),
+            input.getWorkflowExecution(),
+            input.getWorkflowType(),
+            input.getTimeout(),
+            input.getTimeoutUnit(),
+            clientOptions.getDataConverter());
+    return new GetResultAsyncOutput<>(
+        resultValue.thenApply(
+            payloads ->
+                convertResultPayloads(payloads, input.getResultClass(), input.getResultType())));
+  }
+
+  @Override
+  public <R> QueryOutput<R> query(QueryInput<R> input) {
+    WorkflowQuery.Builder query = WorkflowQuery.newBuilder().setQueryType(input.getQueryType());
+    Optional<Payloads> inputArgs =
+        clientOptions.getDataConverter().toPayloads(input.getArguments());
+    inputArgs.ifPresent(query::setQueryArgs);
+    QueryWorkflowRequest request =
+        QueryWorkflowRequest.newBuilder()
+            .setNamespace(clientOptions.getNamespace())
+            .setExecution(
+                WorkflowExecution.newBuilder()
+                    .setWorkflowId(input.getWorkflowExecution().getWorkflowId())
+                    .setRunId(input.getWorkflowExecution().getRunId()))
+            .setQuery(query)
+            .setQueryRejectCondition(clientOptions.getQueryRejectCondition())
+            .build();
+
+    QueryWorkflowResponse result;
+    result = genericClient.query(request);
+
+    boolean queryRejected = result.hasQueryRejected();
+    WorkflowExecutionStatus rejectStatus =
+        queryRejected ? result.getQueryRejected().getStatus() : null;
+    Optional<Payloads> queryResult =
+        result.hasQueryResult() ? Optional.of(result.getQueryResult()) : Optional.empty();
+    R resultValue =
+        convertResultPayloads(queryResult, input.getResultClass(), input.getResultType());
+    return new QueryOutput<>(rejectStatus, resultValue);
+  }
+
+  @Override
+  public CancelOutput cancel(CancelInput input) {
+    // RunId can change if workflow does ContinueAsNew. So we do not set it here and
+    // let the server figure out the current run.
```

Comment:
This is not how I implemented it, it's carried from the previous implementation. WorkflowClientStubImpl saves the workflowExecution after the start call and after that it just reuses it, it doesn't adapt it for each call it makes, so the runId will actually come here right now directly from that initial start in the majority of cases.
I don't think that changing this logic should be a part of this PR. It could potentially be breaking and dangerous and let's keep it as a separate PR for an easier tracing if something goes wrong with this change. I can submit a PR with this change after this PR is merged.